### ISSUE TITLE
Problems Connecting to Server - "Curl failed with code 35 (SSL connect error)"

### ISSUE THREAD

**Issue by Gunmetal-61** [cisco/libacvp#646]
Hello,

I am currently having issues connecting to the server after compiling the program and trying to run it according to the instructions found under the [Running](https://github.com/cisco/libacvp#running) section.

These were the commands I entered from the base working directory (`./libacvp`):
```
export LD_LIBRARY_PATH=/usr/bin/openssl:/usr/bin/curl
. ./scripts/nist_setup.sh
./app/acvp_app --aes
```

This was the output I received.

```
Using the following parameters:

    ACV_SERVER:     demo.acvts.nist.gov
    ACV_PORT:       443
    ACV_URI_PREFIX: /acvp/v1/
    ACV_CA_FILE:    certs/acvp.nist.gov.crt
    ACV_CERT_FILE:  certs/ESVDemo.cer
    ACV_KEY_FILE:   certs/ESVDemo.key

***ACVP [STATUS][acvp_login:2677]--> Logging in...
***ACVP [ERR][acvp_curl_http_post:354]--> Curl failed with code 35 (SSL connect error)

***ACVP [ERR][log_network_status:1245]--> Received no response from server.
***ACVP [STATUS][acvp_login:2689]--> Login Send Failed
***ACVP [ERR][acvp_run:3349]--> Failed to login with ACVP server 
```

To my knowledge, I have specified all parameters in the nist_setup.sh script correctly, my certificates and TOTP seed are valid, and I do not believe it to be a problem with my network connection.  I have also tried using the `mozzila_trust_anchors.pem` file in place of `acvp.nist.gov.crt` in for `ACV_CA_FILE` in nist_setup.sh to no change in outcome or output.

Does anyone have any suggestions on how to solve this issue?  Thank you.

**Comment by abkarcher** [comment]
Hello,

can you try setting the ACV_CA_FILE, ACV_CERT_FILE, etc. to absolute paths? 
Since the TOTP exchange happens within the protocol, this would indicate a problem with the cert/key/CA as opposed to that. Usually I get this error when my ACV_KEY_FILE is not formatted properly or pathed properly. 

Another note looking at the filenames - is your key/cert for the ESV demo server? Those would be separate credentials/URLs than for the ACV demo server.

Thanks,
Andrew

**Comment by Gunmetal-61** [comment]
Hello Andrew,

Thanks for the suggestions.  Looks like the files are being identified properly.  Absolute paths produce no change, but mis-spelling either the relative or absolute path to the files throws a different error evidently related to the inability to find the files if run in that configuration.

You were right about the key/cert accidently being ESV though.  I received a new set from NIST for ACV.  The person at NIST was also very helpful in that he ran my credentials (private key included) on the server side of things and found he could connect properly once the aforementioned issue was rectified + me removing the password protection on the key (which I didn't realize I had).  It's definitely not a server problem or the contents of my key/cert.

Now, when I run `./app/acvp_app --get /acvp/v1/algorithms --verbose`, I get:

```
/libacvp$ ./app/acvp_app --get /acvp/v1/algorithms --verbose

Using the following parameters:

    ACV_SERVER:     demo.acvts.nist.gov
    ACV_PORT:       443
    ACV_URI_PREFIX: /acvp/v1/
    ACV_CA_FILE:    certs/mozzila_trust_anchors.pem
    ACV_CERT_FILE:  certs/Demo.cer
    ACV_KEY_FILE:   certs/Demo.key

***ACVP [INFO][acvp_http_user_agent_handler:1658]--> HTTP User-Agent: libacvp/1.0;Linux;5.8.0-43-generic;x86_64;Intel(R) Core(TM) i5-7440HQ CPU @ 2.80GHz;GCC/9.3.0

***ACVP [STATUS][acvp_login:2677]--> Logging in...
***ACVP [ERR][acvp_curl_http_post:354]--> Curl failed with code 60 (SSL peer certificate or SSH remote key was not OK)

***ACVP [INFO][log_network_status:1215]--> POST Login...
        Status: 0
        Url: https://demo.acvts.nist.gov:443/acvp/v1/login
        Resp: Recieved

***ACVP [ERR][log_network_status:1245]--> Received no response from server.
***ACVP [STATUS][acvp_login:2689]--> Login Send Failed
***ACVP [ERR][acvp_run:3349]--> Failed to login with ACVP server
```

With the verbose output (and all along the way before the key/cert content was proofread), the verbose output showing the `POST Login` request to the server has indicated it has received my request, though the NIST staff member skimmed over the logs and said he did not see any activity in the logs from me at a glance.

I believe I also compiled the program correctly according to the [README Build section (https://github.com/cisco/libacvp#building).  These were my commands:

```
./configure --with-ssl-dir=/usr/lib/x86_64-linux-gnu --with-libcurl-dir=/usr/lib/x86_64-linux-gnu
make clean
make
sudo make install
```

I'm running Ubuntu 20.04 LTS in a VirtualBox VM on Windows 10, and SSH'ing into the VM from Windows with the VS Code SSH extension for ease of development.

**Comment by abkarcher** [comment]
Hello,

Thanks for all of the info! That's extremely helpful. 

It looks like we have made some progress; curl code 60 indicates a problem verifying the remote server now. The http return code 0 means that no login was initiated.

I have not seen curl code 60 from libacvp before, but it seems likely this is a problem with the CA. Have you tried running without setting ACV_CA_FILE? acvp login seems to work in my Ubuntu VM without it. Or, maybe try using a different or newer CA file? You can acquire a newer one here - https://curl.se/docs/caextract.html (We will look into updating ours if this turns out to be an issue).

Thanks,
Andrew

**Comment by Gunmetal-61** [comment]
Hello Andrew,

Apologies for the late reply, but we did discover the root of the issue after getting a working set of keys.  It was not an issue with the NIST servers, libacvp, or even my OS config.  I ended up copying my Ubuntu VM image over to a second machine without changing anything within and managed to successfully connect on first try.  Turns out our internal company network firewall was blocking traffic for whatever reason, but the guest network in the very same office (which the second machine was connected to) allowed the request to go through properly.

Consider this issue closed!  Thanks for all the help.
